src/track/globaltrackcache.h

```diff
@@ -208,35 +246,50 @@ class GlobalTrackCache {
     // to verify the class invariants during development.
     bool verifyConsistency() const;
 
-    TrackPointer lookupInternal(
-            const TrackId& trackId) const;
-    TrackPointer lookupInternal(
-            const TrackRef& trackRef) const;
+    TrackRefPtr lookupByIdInternal(
+            const TrackId& trackId);
+    TrackRefPtr lookupByRefInternal(
+            const TrackRef& trackRef);
+
+    TrackRefPtr reviveInternal(
+            const Item& item);
 
     bool resolveInternal(
             GlobalTrackCacheResolver* pCacheResolver,
             TrackRef* pTrackRef,
             const TrackId& trackId,
             const QFileInfo& fileInfo);
 
-    TrackRef updateTrackIdInternal(
-            const TrackPointer& pTrack,
+    TrackRef initTrackIdInternal(
+            TrackPointer pTrack,
             const TrackRef& trackRef,
             TrackId trackId);
 
-    Item purgeInternal(
-            const TrackRef& trackRef);
-
-    void evict(
+    bool evictAndDelete(
             Track* pTrack);
-    Track* evictInternal(
+
+    typedef std::set<Track*> AllocatedTracks;
+
+    bool evictAndDeleteInternal(
             GlobalTrackCacheLocker* /*nullable*/ pCacheLocker,
-            const TrackRef& trackRef);
+            AllocatedTracks::iterator ipIndexedTrack,
+            bool evictUnexpired);
+    bool evictInternal(
+            const TrackRef& trackRef,
+            AllocatedTracks::iterator ipIndexedTrack,
+            bool evictUnexpired);
+
+    void afterEvicted(
+            GlobalTrackCacheLocker* /*nullable*/ pCacheLocker,
+            Track* pEvictedTrack);
 
     GlobalTrackCacheEvictor* m_pEvictor;
 
     mutable QMutex m_mutex;
 
+    AllocatedTracks m_indexedTracks;
```

Comment:
This should be a set of weak pointers. If we do not store plain pointers at all, we should have no pain with race conditions, except stray qt signals.

Reply:
Nope. See my comment about weak/plain pointers.

Reply:
GitHub seems to hide you referenced comment. Can you put a link to it?